specs/chugsplash-manager.md

```diff
@@ -0,0 +1,165 @@
+# ChugSplashManager Speed Spec
+
+## Bundle states
+
+```typescript
+enum ActionType {
+  SET_CODE,
+  SET_STORAGE
+}
+```
+
+```typescript
+enum BundleStatus {
+  EMPTY,
+  PROPOSED,
+  APPROVED,
+  CANCELLED,
+  COMPLETED
+}
+```
+
+```typescript
+struct BundleState {
+  BundleStatus status;
+  bool[] executions;
+  uint256 actionsExecuted;
+  uint256 timeClaimed;
+  address selectedExecutor;
+}
+```
+
+```typescript
+struct Action {
+  string target;
+  ActionType actionType;
+  bytes code;
+}
+```
+
+## Computing a unique bundle ID
+
+```typescript
+const computeBundleId = (
+  bundleRoot: bytes32,
+  bundleSize: uint256,
+  configUri: string
+): bytes => {
+  return keccak256(
+    abi.encode(
+      bundleRoot,
+      bundleSize,
+      configUri
+    )
+  )
+}
+```
+
+## Getting the selected executor
+
+```typescript
+const getSelectedExecutor = (
+  bundleId: bytes32
+): address => {
+  return bundles[bundleId].selectedExecutor
+}
+```
+
+## Getting the proxy by name
+
+```typescript
+const getProxyByName = (
+  name: string
+): address payable => {
+  return (
+    payable(
+      Create2.compute(address(this), keccak256(bytes(name)), type(Proxy).creationCode)
+    )
+  )
+}
+```
+
+## Proposing a ChugSplash bundle
+
+* A ChugSplash bundle MUST only be proposable by the `admin` of the `ChugSplashManager`.
+* Bundles MUST be in the `EMPTY` state before they can be proposed.
+* Proposing a bundle puts the bundle in the `PROPOSED` state.
+* The ChugSplashManager MUST emit the event ChugSplashBundleProposed with the ID of the bundle, the bundle root, the bundle size, and the config URI as parameters. The ChugSplashManager MUST announce this event to the registry.
+
+## Approving a ChugSplash bundle
+
+* A ChugSplash bundle MUST only be approvable by the `admin` of the `ChugSplashManager`.
+* Bundles MUST have be in the `PROPOSED` state before they can be approved.
+* There MUST NOT be any active bundle.
+* Approving a bundle MUST put the bundle in the `APPROVED` state.
+* Approving a bundle MUST make the approved bundle the active bundle.
+* The balance of the ChugSplashManager minus the current totalDebt MUST be greater than BUNDLE_EXECUTION_BOND_AMOUNT ETH in its balance.
```

Comment:
We may want an additional buffer here to account for gas price variability.  Having enough to cover the bond, but not send it back, could be a problem.